sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -135,9 +135,15 @@
  *     .apply(FileIO.readMatches().withCompression(GZIP))
  *     .apply(MapElements
  *         // uses imports from TypeDescriptors
- *         .into(KVs(strings(), strings()))
- *         .via((ReadableFile f) -> KV.of(
- *             f.getMetadata().resourceId().toString(), f.readFullyAsUTF8String())));
+ *         .into(kvs(strings(), strings()))
+ *         .via((ReadableFile f) -> {
+ *           try {
+ *             return KV.of(
+ *                 f.getMetadata().resourceId().toString(), f.readFullyAsUTF8String());
+ *           } catch (IOException ex) {
+ *             throw new RuntimeException("Failed to read the file", ex);
```

Comment:
Is there a better idea than this exception handling?

Reply:
`via()` takes a `Contextful<Fn<..>>`, where `Fn.apply()` is allowed to throw `Exception` (https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Contextful.java#L70) - so the original snippet should compile, no?

Reply:
Interesting; is there a way to use lambda while letting a checked exception thrown?

My IntelliJ complains as below:

<img width="697" alt="Screen Shot 2019-12-03 at 09 07 42" src="https://user-images.githubusercontent.com/28604/70058271-96fde000-15ac-11ea-9ea0-fb3057974bdb.png">


Also from Maven:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project word-count-beam: Compilation failure
[ERROR] /Users/suztomo/word-count-beam/src/main/java/org/apache/beam/examples/WordCount.java:[225,85] unreported exception java.io.IOException; must be caught or declared to be thrown
```

IntelliJ says it picks up this method (not the one with `Contextful`):

```
  /** Binary compatibility adapter for {@link #via(ProcessFunction)}. */
  public <NewInputT> MapElements<NewInputT, OutputT> via(
      SerializableFunction<NewInputT, OutputT> fn) {
    return via((ProcessFunction<NewInputT, OutputT>) fn);
  }
```

Reply:
Ah, then maybe the example should add `Requirements.empty()` to force this overload? https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Contextful.java#L122 There's probably some frustrating Java reason why we can't just add an `fn(Fn)` overload - Java probably wouldn't be able to disambiguate between it and the other overloads when you supply a lambda.

Reply:
IntelliJ says: `Cannot resolve method 'via(<lambda expression>, org.apache.beam.sdk.transforms.Requirements)'`

![image](https://user-images.githubusercontent.com/28604/70067467-51491380-15bc-11ea-957c-eb15b1d76f10.png)

This worked. Is this what you're looking for:

```
        .apply(MapElements.into(kvs(strings(), strings()))
            .via(Contextful.fn(
                (ReadableFile f, Context c) -> KV.of(f.getMetadata().resourceId().toString(), f.readFullyAsUTF8String()),
                Requirements.empty()
            )))
```

![image](https://user-images.githubusercontent.com/28604/70068344-d84abb80-15bd-11ea-88f9-bcec93ee12dc.png)

Reply:
@jkff Did you get a chance to check the code above?

Reply:
Sorry I missed this. Hmm, this is getting ugly enough that I think the original code you proposed is better - it's, at least, idiomatic in Java. Would be nice (not in this PR) to make it easier to do MapElements.via() with throwing functions without the hassle of fn(lambda, Requirements.empty()).

So for now I suggest to revert to the original version you proposed.

Reply:
Created a ticket to followup https://issues.apache.org/jira/browse/BEAM-8942 .